SWIPs/swip-adaptive-node-protocol.md

```diff
@@ -0,0 +1,131 @@
+
+## Specification
+
+Upon initial connection, capabilities can be transmitted already within the handshake message.
+
+Capabilities are combinational. Some combination may make more sense than other. Interpreting which combinations are feasible and which are not should be handled by respective nodes, and not be hardwired in the protocol.
+
+For this reason, a bit vector seems the most logical choice. The bit vector can represent the adaptive node operation state within the node, and be embedded into the protocol message as-is.
+Since the capabilities may have to be set from different system modules, the network module should provide an API for registering them. The following structure is proposed:
+
+* the first byte is the identifier of the module (0x00 is bzz, and built in)
+* the second byte is the length `n` of the bit vector containing the flags
+* the last `n` bytes is the bit vector containing the flags
+
+For the individually capability flags we will initially be choosing from, we can see an example of two distinct byte pairs representing tematic categories:
+
+```
+BZZ/SYNC: 			// (registered by default)
+- retrieve		0x00020001
+- push			0x00020002
+- retrieve relay	0x00020010
+- push sync relay	0x00020020
+- storer node		0x00028000	// the node participates as storer for area of responsibility
+
+PSS:
+- receive	0x010101
+- send		0x010102
+- relay		0x0l0110
+```
+
+In the golang implementation a bitvector of capabilities will be represented as a key-value pair:
+
+type Capability struct {
+	Id  CapabilityID
+	Cap []bool
+}
+
+We define a new protocol message for the `bzz` protocol called `Capabilities`. This message contains one field holding the individual module bitvectors. This protocol message can be sent at any time to inform peers of _changes_ to the node's capabilities. 
+
+```
+Capabilities {
+	Caps []Capability
+}
+```
+
+_Note that the RLP serialization of this structure in the golang implementation adds an extra list entry for every struct encapsulation. This "list of capabilities" will in reality be a "list of list of capabilities," where the former list only has one element._
+
+Furthermore, the bzz handshake message itself should also embed this message instead of the `Light` field that is currently there. Incidentally, in the golang implementation, the `BzzAddr` object transmitted in the bzz handshake is also (part of) the peer record stored by the kademlia peer database. A practical approach is therefore to embed the `Capabilities` information in the same structure. `BzzAddr` thus becomes:
+
+type BzzAddr struct {
+	OAddr        []byte
+	UAddr        []byte
+	Capabilities *Capabilities
+}
+
+Consequently the `LightNode` flag is removed from the `HandshakeMsg` message.
+
+```
+HandshakeMsg {
+	Version		uint64
+	NetworkID	uint64
+	Addr		*BzzAddr
+}
+```
+
+For simplicity, arrays of bools are used to represent the flags in interfacing code. In the JSON-RPC API a `Capability` is represented in the following manner:
```

Comment:
implementation detail, perhaps add to implementation suggestion section to keep spec more concise?